### ISSUE TITLE
[DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 

### ISSUE THREAD

**Issue by Boehmer86** [nicoduj/homebridge-deebotEcovacs#51]
Hello all,

I am not sure how I caused it but right now the deebotEcovacs plugin always makes my Homebridge crash. I uninstalled and removed the config and accessories, upgraded all (nodes, home bridge, etc.) and then re installed all to no avail. 

After startup the plugin tries to connect every split second and fails, thus pilling up lots of uncleared memory. 
The system memory is then shrinking by the second, you can see the counter sinking fast until it crashes. 

When I turn on debug immediately after startup this comes up spamming the log:
```
...
[2/10/2021, 2:01:23 AM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[2/10/2021, 2:01:23 AM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[2/10/2021, 2:01:23 AM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout
...
```

For reference my config:
`
        {
            "email": "____",
            "password": "_____",
            "countryCode": "DE",
            "deebotNames": [
                "Ozmo"
            ],
            "refreshTimer": 300,
            "publishBipSwitch": true,
            "publishSwitch": true,
            "publishFan": false,
            "publishMotionDetector": false,
            "publishAutoSwitch": true,
            "publishEdgeSwitch": false,
            "publishSpotSwitch": true,
            "publishSpotAreaSwitches": [
                "0",
                "1",
                "2",
                "3",
                "4",
                "5",
                "6",
                "7",
                "8",
                "9",
                "10",
                "11",
                "12",
                "13",
                "14"
            ],
            "platform": "DeebotEcovacs"
        }
`

I am grateful for any hint or suggestion.

Note, I tripple checked the Pi connectivity, all works fine, DNS is healthy, Homebridge and all other plugins work and connect to the great wide world as well as my home network happily. I also do not have any packet loss problems. 

With greetings

Robert

**Comment by nicoduj** [comment]
Hi,

it seems to retrieve the config from ecovacs server, but can't connect to your deebot. Did you check with ecovacs app ? did you try without the deebot name just in case ?

**Comment by Boehmer86** [comment]
Good Morning @nicoduj 

I tried removing the name but the result is the same, however I now captured the full log output from a fresh start-up which shows that the connections initially seems to work (or am I wrong?):
`[2/10/2021, 10:41:48 AM] [DeebotEcovacs] INFO - connected
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - getDeebots : [{"did":"42eeae5f-1954-44a5-9039-__________","name":"E08M14103________","class":"fqxoiu","resource":"Cc3g","nick":"Ozmo","company":"eco-ng","bindTs":1609604018847,"service":{"jmq":"jmq-ngiot-eu.dc.ww.ecouser.net","mqs":"api-ngiot.dc-as.ww.ecouser.net"}}]
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : false
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : false
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - SpotArea Cleaning for Ozmo : true
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : true
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating Battery Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating Main Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating Sound stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating Auto stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 0 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 1 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 2 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 3 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 4 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 5 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 6 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 7 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 8 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 9 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 10 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 11 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 12 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 13 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Creating SpotArea 14 stateless Switch Service Ozmo
[2/10/2021, 10:41:49 AM] [DeebotEcovacs] INFO - Setting Timer for background refresh every  : 250s
[2/10/2021, 10:41:50 AM] [DeebotEcovacs] INFO - Vacbot Ozmo ready: "Client connected. Subscribe successful"
[2/10/2021, 10:42:05 AM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[2/10/2021, 10:42:05 AM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[2/10/2021, 10:42:06 AM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout `

With greetings 

Robert

**Comment by nicoduj** [comment]
I think you can connect to ecovacs server, retrieve the conf of your deebot, but communication with your osmo seems broken. The ecovacs app is ok ? You can try to remove the refreshTimer , and tell me if any command works.

**Comment by Boehmer86** [comment]
Hi @nicoduj 

Very interesting, thanks a lot for the hint!

I checked the Ecovacs App and noticed the robot is offline. It was online yesterday evening. I used the app for some manual cleaning. 

You are correct, I changed my network setup, actually quite drastic. 
I enabled WPA3 for my main network and changed the password. I created a MAC locked WPA2-AES IoT WiFi for all my HomeKit IoT devices. I used the WiFi reset button and the Ecovacs App to connect the robot to the new WiFi and I thought it worked. However I am now no longer 100% sure. Right now, I did a full reset of the Deebot and configured it for the new WiFi. Against the clear recommendation of Ecovacs I assigned it a static IP via a DHCP reservation. The reason I did so is because I see absolutely no logical reason on why this should be an issue.

Now after this reset, I re-enabled my original configuration which worked for several weeks now (see my first post) and this is the current debug startup output:
`[2/10/2021, 12:27:23 PM] [DeebotEcovacs] INFO - connected
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - getDeebots : [{"did":"42eeae5f-1954-44a5-9039-______","name":"E08M14103______","class":"fqxoiu","resource":"Cc3g","nick":"Ozmo","company":"eco-ng","bindTs":1609604018847,"service":{"jmq":"jmq-ngiot-eu.dc.ww.ecouser.net","mqs":"api-ngiot.dc-as.ww.ecouser.net"}}]
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : false
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : false
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - SpotArea Cleaning for Ozmo : true
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : true
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating Battery Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating Main Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating Sound stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating Auto stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 0 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 1 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 2 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 3 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 4 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 5 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 6 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 7 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 8 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 9 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 10 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 11 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 12 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 13 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Creating SpotArea 14 stateless Switch Service Ozmo
[2/10/2021, 12:27:24 PM] [DeebotEcovacs] INFO - Setting Timer for background refresh every  : 250s
[2/10/2021, 12:27:25 PM] [DeebotEcovacs] INFO - Vacbot Ozmo ready: "Client connected. Subscribe successful"
[2/10/2021, 12:27:26 PM] [DeebotEcovacs] INFO - Charge status for Ozmo : charging false true
[2/10/2021, 12:27:26 PM] [DeebotEcovacs] INFO - Clean status for Ozmo : idle
[2/10/2021, 12:27:26 PM] [DeebotEcovacs] INFO - Charge status for Ozmo : charging false true
[2/10/2021, 12:27:26 PM] [DeebotEcovacs] INFO - Battery level for Ozmo : 100 100 0
[2/10/2021, 12:27:26 PM] [DeebotEcovacs] INFO - Charge status for Ozmo : charging false true`

I will monitor this now for the next hours and days and will close the ticket by tomorrow if it continues to work. It likely was "my fault". However just noting, there should be some protection against the plugin eating all system memory in such a "loss of connectivity" scenario. It is kind of a race condition. 

With many greetings

Robert

**Comment by nicoduj** [comment]
Hi, thanks for coming back. You are right for the race condition, will try to find time to reproduce, it should not eat memory at all .

**Comment by Boehmer86** [comment]
Hi @nicoduj 

To add, the problem occurred again and I found two remarks in the Netgear Orbi support community which helped resolve the connectivity issue:

1.) After applying a DHCP reservation the router needs to be rebooted (seriously!) else the DHCP server does not renew but also not supply the new address. I am not sure this caused the issue but it matches some weird DHCP log entries for the Deebot that came up after I inserted the reservation before rebooting.

2.) I disabled "20/40 Mhz co-existence" for the 2.4 GHz WiFi as this was flagged as not reliable both in Deebot and NETGEAR Communities. 

In regards to the Deebot Ecovacs plugin I disabled the refresh interval as suggested by you @nicoduj and the result towards the race condition changed. When the bot dropped off the WiFi (before I made the two above mentioned changes) the system would start complaining and loosing memory but only very little, much less as before. It almost looked like "stable", also there where only 5-6 messages in the log about the connectivity issue and not as before like several hundred. 

I did not check the codebase but my guess would be that the refresh timer spawns some update collection process which goes into Limbo if the Deebot connection is lost.

I hope this helps.

Should I close the ticket now or do you want to leave it open until the race condition is found?

Best

Robert

**Comment by nicoduj** [comment]
Thank you, I will keep this open for the moment.

Concerning orbi yes they are tricky ... disabling 20/40 coexistence definitely improved connection here also, and I don't use it as a router / dhcp server, only as a wifi access point :)

**Comment by Boehmer86** [comment]
#offtopic 

Thanks @nicoduj for the note and confirming the 20/40 2.4 GHz topic for Orbi, great too get support here even in this matter ;-) I made the switch to the Netgear Orbi Router mode out of necessity but I am regretting it. My Fritz!Box failed with a hardware fault and the replacement from my ISP was an older model which could not load any of my configurations so I had to start all over anyway. However it is frustrating to see how poorly Netgear handles even simple operating system tasks on their router. There should be a legal initiative demanding that open firmware become a legal requirement for sole device manufacturers. It reduces their liability once they go bankrupt and it allows the community to enhance and fix things for current and old units. 

Best, Robert

**Comment by nicoduj** [comment]
to continue #offtopic, I think there is a way to put Voxel custom firmware to some ORbi system, but never tried since like I said I use them only as access points and never found a better solution in my configuration for wifi range/ perf (even tried some unifi without any success) https://www.reddit.com/r/orbi/comments/gs6loq/custom_firmware_build_for_orbi_rbk50rbk53_rbr50/

**Comment by Boehmer86** [comment]
Thanks a lot for the hint with the firmware @nicoduj 

Just for anyone ending up here using Google:

- Disabling 20/40Mhz co-existence for 2.4GHz did not fully solve the problem
- I needed to disable the DHCP function of the Netgear Orbi and user another DHCP Server, in my case hosted on a Raspberry

Context:
After disabling the co-existence feature my Deebot did not disconnect every 1-2 hours from the WiFi but exactly after 24 hrs. This immediately made me aware of the DHCP configuration of my Netgear Orbi which grants leases for 24 hrs. After debugging I found that the Deebot does some weird network voodoo and enters into a state where he, after 24 hours constantly floods the Netgear with DHCP requests but does not accept the response and just keeps asking. It is still reachable until the actual expiration but it already starts flooding after half the lease time is expired (the renewal interval). 

On the internet I found that the Netgear DHCP implementation is incredibly poor and it only add to my experience with other devices. I have an Eve HomeKit Camera with similar issues. So I configured my PiHole to act as the new DHCP server. This seem to have solved the problem for now both for the Eve Cam and Deebot. However I will have to check and wait for a full lease cycle.

However, to note, disabling the WiFi co-existence made the WiFi connection stable. The next issue was then the DHCP on my network due to the Netgear Orbi. All other issues are simply a result of the mentioned race-condition. Once the Deebot drops of the net, my homebridge starts venting memory (much slower if I disable the refresh timer but enough to crash after a few hours).
